lib/ttfunk/table/cff/dict.rb

```diff
@@ -0,0 +1,225 @@
+              result << encode_operator(operator)
+            end
+          end
+        end
+
+        private
+
+        def encode_operator(operator)
+          if operator >= 1200
+            [12, operator - 1200].pack('C*')
+          else
+            [operator].pack('C')
+          end
+        end
+
+        def encode_operand(operand)
+          bytes = case operand
+                  when Integer
+                    encode_integer(operand)
+                  when Float, BigDecimal
+                    encode_float(operand)
+                  when Sci
+                    encode_sci(operand)
+                  end
+
+          bytes.pack('C*')
+        end
+
+        def encode_integer(int)
+          case int
+          when -107..107
+            [int + 139]
+
+          when 108..1131
+            int -= 108
+            [(int >> 8) + 247, int & 0xFF]
+
+          when -1131..-108
+            int = -int - 108
+            [(int >> 8) + 251, int & 0xFF]
+
+          when -32_768..32_767
+            [28, (int >> 8) & 0xFF, int & 0xFF]
+
+          else
+            encode_integer32(int)
+          end
+        end
+
+        def encode_integer32(int)
+          [
+            29,
+            (int >> 24) & 0xFF,
+            (int >> 16) & 0xFF,
+            (int >> 8) & 0xFF,
+            int & 0xFF
+          ]
+        end
+
+        def encode_float(float)
+          pack_decimal_nibbles(encode_significand(float))
+        end
+
+        def encode_sci(sci)
+          sig_bytes = encode_significand(sci.significand)
+          exp_bytes = encode_exponent(sci.exponent)
+          pack_decimal_nibbles(sig_bytes + exp_bytes)
+        end
+
+        def encode_exponent(exp)
+          return [] if exp == 0
+          [exp > 0 ? 0xB : 0xC, *encode_significand(exp)]
+        end
+
+        def encode_significand(sig)
+          sig.to_s.each_char.with_object([]) do |char, ret|
+            case char
+            when '0'..'9'
+              ret << char.to_i
+            when '.'
+              ret << 0xA
+            when '-'
+              ret << 0xE
+            else
+              break ret
+            end
+          end
+        end
+
+        def pack_decimal_nibbles(nibbles)
+          packed = [30]
+
+          nibbles.each_slice(2).each do |(high_nb, low_nb)|
+            # low_nb can be nil if nibbles contains an odd number of elements
+            low_nb ||= 0xF
+            packed << (high_nb << 4 | low_nb)
+          end
+
+          packed << 0xFF if nibbles.size.even?
+          packed
+        end
+
+        def parse!
+          @dict = {}
+          operands = []
+
+          # @length must be set via the constructor
+          while io.pos < table_offset + length
+            case b_zero = read(1, 'C').first
+            when 12
+              operator = decode_two_byte_operator
+              @dict[operator] = operands
+              operands = []
+            when 0..21
+              @dict[b_zero] = operands
```

Comment:
It probably may prove prudent to validate data type of operands for a specific operator.

Reply:
Would you like to see that added before merge? If so I can work something up.

Reply:
Yes, please. It seems like a logical place to add it if at all.

Reply:
Ok, I realized the problem is that dicts themselves don't really have the concept of valid or invalid data - they're just data structures. Individual types of dict (like the top dict, private dict, etc) do contain specific operators and operands, but I think it makes sense to add validation when those particular classes are introduced. Thoughts?

Reply:
I don't know. Operators are present in the parsing logic. Even though operators are not universal (used only in specific dicts) they have specific types of operands. In my mind it makes sense to validate operands here. Maybe add additional validations to specific dicts that restrict operators. But that doesn't seem to be connected with operand validations.